### PR TITLE
Add BigQuery hook.

### PR DESCRIPTION
**Pull request by criccomini**
I've added support for [Google BigQuery](https://cloud.google.com/bigquery/) hooks. This allows us to connect to BQ via a hook, as well as query BigQuery from the `ad hoc` page. It also enables charts to generate based off of a BigQuery query.

### CODE REVIEW

**Comment by criccomini** [comment]
Note I haven't written any tests for this. I did validate that it works for our BigQuery install. Was able to create both charts and ad hoc queries with it.

File: airflow/hooks/bigquery_hook.py [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+        if len(response['rows']) > 0:
+            # Extract column names from response.
+            columns = [c['name'] for c in response['schema']['fields']]
+
+            # Extract data into a two-dimensional list of values.
+            data = []
+            for row in response['rows']:
+                row = map(lambda kv: kv['v'], row['f'])
+                data.append(row)
```
Comment (max-sixty) [inline airflow/hooks/bigquery_hook.py]: I don't know much about GBQ - but in general this approach of iterating through things in python is going to be a few orders of magnitude slower than delegating to pandas. Did you try using this: http://pandas.pydata.org/pandas-docs/stable/generated/pandas.io.gbq.read_gbq.html?

Comment (criccomini) [inline airflow/hooks/bigquery_hook.py]: Yea, I had it on my TODO. This kind of scared me away though:

```
THIS IS AN EXPERIMENTAL LIBRARY
```

Comment (criccomini) [inline airflow/hooks/bigquery_hook.py]: Poked at this a bit. The Pandas GBQ implementation appears to use 3 legged oauth, and thus requires a user at the terminal to actually be present. Running gbq.read_gbq actually opens a browser window for the human to complete OAuth. This doesn't work with two legged OAuth, which requires a service account to run. My entire implementation is predicated off of two legged OAuth. Some discussion on Google cloud is documented [here](https://developers.google.com/identity/protocols/OAuth2ServiceAccount).

Comment (criccomini) [inline airflow/hooks/bigquery_hook.py]: https://github.com/pydata/pandas/blob/master/pandas/io/gbq.py#L199 Seems to be the most interesting part of Pandas' GBQ integration.

Comment (criccomini) [inline airflow/hooks/bigquery_hook.py]: Figured out a hack. PR update incoming.

**Comment by criccomini** [comment]
Latest PR uses pandas.io.gbq to create DataFrames. This allows Airflow/GBQ integration to support pagination, types in schemas, and long-running queries, among other things.

**Comment by criccomini** [comment]
I have a tweaked implementation that just uses BaseHook. It's unclear to me the relation between BaseHook and DbApiHook. I can make both work with the BQ hook, but it seems like BaseHook is actually the correct one. Is this understanding right?

**Comment by mistercrunch** [comment]
Hey can we move this in `airflow.contrib.hooks` to make it clear that it isn't unit tested and testable by the maintainers?

File: airflow/hooks/__init__.py [hunk 2]
```diff
@@ -13,6 +13,7 @@
     'webhdfs_hook': ['WebHDFSHook'],
     'mysql_hook': ['MySqlHook'],
     'postgres_hook': ['PostgresHook'],
+    'bigquery_hook': ['BigQueryHook'],
```
Comment (mistercrunch) [inline airflow/hooks/__init__.py]: this will need to move to airflow.contrib.hooks.**init**.py

**Comment by mistercrunch** [comment]
Please add an entry in `docs/code.rst`

**Comment by mistercrunch** [comment]
DbApiHook provides tons of methods provided that the database API follows the python spec defined here:
https://www.python.org/dev/peps/pep-0249/

All of the db drivers implementing that standard can share a lot of code.

**Comment by criccomini** [comment]
Will do

**Comment by criccomini** [comment]
> All of the db drivers implementing that standard can share a lot of code.

Ah, got it. The BQ connection does not implement this. I suppose we could wrap the BQ connection in our own custom PEP-0249 implementation, but I'll leave that for a separate ticket.

**Comment by criccomini** [comment]
@mistercrunch I made the change to move the hook to contrib locally, but it makes the ad hoq connection disappear. The models.py/Connection.get_hook currently points to:

```
elif self.conn_type == 'bigquery':
  return hooks.BigQueryHook(bigquery_conn_id=self.conn_id)
```

I could obviously update this to point to the contrib hooks directory instead, but this feels wrong. I see there's some plugin_manager stuff in the init py file. What do you recommend?

**Comment by criccomini** [comment]
Note: I see this for the Vertica hook as well, which is in contrib:

```
'module' object has no attribute 'VerticaHook'
```

Perhaps I need to enable contribs somehow?

**Comment by mistercrunch** [comment]
`return contrib.hooks.BigQueryHook(bigquery_conn_id=self.conn_id)` is fine by me

**Comment by criccomini** [comment]
@mistercrunch Added code.rst update. Imported contrib.hooks in models.py. Please have another look.

**Comment by criccomini** [comment]
Any update here?

**Comment by criccomini** [comment]
@mistercrunch any update? This has been pending for 11 days.

**Comment by mistercrunch** [comment]
Sorry about the delay, many of us took a nice break over the holidays and I wanted to stabilize a branch before letting a wave of stuff in. Reviewing now :)

**Comment by mistercrunch** [comment]
:+1: looks solid, merging. Thanks for your patience :)

**Pull request merged**